google/cloud/security/common/gcp_type/firewall_rule.py

```diff
@@ -65,11 +82,166 @@ def __init__(self, **kwargs):
+                            list(self._destination_ranges)),
+                           ('priority', self._priority)]:
+            if value:
+                firewall_dict[key] = value
+        return json.dumps(firewall_dict, sort_keys=True)
+
+    def validate(self):
+        """Validates that a rule is valid and not a duplicate.
+
+        Validation is based on reference:
+        https://cloud.google.com/compute/docs/reference/beta/firewalls and
+        https://cloud.google.com/compute/docs/vpc/firewalls#gcp_firewall_rule_summary_table
+        If add_rule_callback is set, this will also confirm that
+        add_rule_callback returns True for the rule, otherwise it will not add
+        the rule.
+
+        Returns:
+          bool: If rule is valid.
+
+        Raises:
+          DuplicateFirewallRuleNameError: Two or more rules have the same name.
+          InvalidFirewallRuleError: One or more rules failed validation.
+        """
+        self._validate_keys()
+        self._validate_direction()
+        self._validate_priority()
+        if not self.firewall_action:
+            raise InvalidFirewallRuleError('Rule missing action "%s"' % self)
+        else:
+            self.firewall_action.validate()
+
+        # TODO: Verify rule name matches regex of allowed
+        # names from reference
+
+        return True
+
+    def _validate_keys(self):
+        """Checks that required keys and value restrictions.
+
+        Required fields: name and network
+        Length restrictions:
+          * name <= 63 characters
+          * <= 256 values:
+            sourceRanges, sourceTags, targetTags, destinationRanges
+
+        Raises:
+          InvalidFirewallRuleError: If keys don't meet requirements.
+        """
+        if not self.name:
+            raise InvalidFirewallRuleError(
+                'Rule missing required field "%s"' % 'name')
+        if not self.network:
+            raise InvalidFirewallRuleError(
+                'Rule missing required field "%s"' % 'network')
+
+        if len(self.name) > 63:
+            raise InvalidFirewallRuleError(
+                'Rule name exceeds length limit of 63 chars: "%s".' %
+                self.name)
+
+        max_256_value_keys = [
+            ('sourceRanges', self._source_ranges),
+            ('sourceTags', self._source_tags),
+            ('targetTags', self.target_tags),
+            ('destinationRanges', self._destination_ranges)
+        ]
+        for key, value in max_256_value_keys:
+            if value and len(value) > 256:
+                raise InvalidFirewallRuleError(
+                    'Rule entry "%s" must contain 256 or fewer values: "%s".'
+                    % (key, value))
+
+    def _validate_direction(self):
+        """Checks that the direction and associated fields are valid.
+
+        Raises:
+          InvalidFirewallRuleError: If:
+            * Direction is 'ingress' and
+              * there are no source ranges or tags
```

Comment:
service accounts should follow the same rules as tags:

ingress: there must be one or more source ranges, tags or service accounts.
ingress: there must not be **any** destination ranges.
ingress: either target tags or service accounts are allowed, but optional.

egress: there must not be **any** source ranges, tags or service accounts.
egress: there may be one or more destination ranges, if no range is specified it defaults to 0.0.0.0/0
egress: either target tags or target service accounts are allowed, but optional.

Reply:
Done.